File: MoonEngine-Root/Kernel/src/main/java/com/greentree/engine/moon/kernel/ReadOnlyClassProperties.java
package com.greentree.engine.moon.kernel;

import java.util.Optional;

public interface ReadOnlyClassProperties<P> extends Iterable<P> {

    default <T extends P> T get(Class<? extends T> cls) {
        Optional<T> property;
        try {
            property = getProperty(cls);
        } catch (Exception e) {
            throw new IllegalArgumentException("get property of class: $cls", e);
        }
        return property.orElseThrow(() -> new IllegalArgumentException("not found property of class: $cls"));
    }

    <T extends P> Optional<T> getProperty(Class<? extends T> cls);

    default boolean contains(Class<? extends P> cls) {
        return getProperty(cls).isPresent();
    }

}

File: MoonEngine-Root/Kernel/src/main/java/com/greentree/engine/moon/kernel/ReadOnlyClassProperties.kt
package com.greentree.engine.moon.kernel

import java.util.*

interface ReadOnlyClassProperties<P> : Iterable<P> {

	operator fun <T : P> get(cls: Class<out T>): T {
		val property = try {
			getProperty(cls)
		} catch(e: Exception) {
			throw IllegalArgumentException("get property of class: \$cls", e)
		}
		return property.orElseThrow { IllegalArgumentException("not found property of class: \$cls") }
	}

	fun <T : P> getProperty(cls: Class<out T>): Optional<T>
	operator fun contains(cls: Class<out P>): Boolean {
		return getProperty(cls).isPresent
	}
}